Add code review standards

djbingham opened the pull request:
https://eaflood.atlassian.net/browse/PSD-216

Commit:
Add code review standards

https://eaflood.atlassian.net/browse/PSD-216

johnwatson484 commented:
I would also add a link from the README page to this so it can easily be navigated to.

johnwatson484 commented on docs/code-review.md:
```diff
@@ -0,0 +1,97 @@
+
+Code reviews should focus on what is being changed and whether the change is appropriate.
+
+Expanding the scope of a pull request at the review stage is not acceptable. It is generally more valuable to swiftly conclude a piece of work that the team has prioritised than to opportunistically seek addition changes, such as refactoring related code. That said, it can be useful to comment on refactoring opportunities without blocking the pull request. It may be that the author has some spare ...
```
slight typo - "addition"

djbingham replied:
Thanks, corrected.

johnwatson484 commented on docs/code-review.md:
```diff
@@ -0,0 +1,97 @@
+Is all new code extensible and easy for other developers to understand? Does it follow common design patterns? Look for unnecessary complexity and remember that this is subjective so take care not to be overly critical or opinionated when commenting on maintainability. Try to make specific suggestions for improvement, rather than simply stating a problem, but use open-ended language to encourage  ...
+
+### Duplication
+Is there duplication within new code or between new and existing code? Could an existing abstraction be reused or should a new abstraction be created? Consider using non-blocking comments when proposing abstractions so they can be addressed separately as technical debt, rather than extending the scope of the pull request.
```
I see pull requests as an opportunity to reduce the amount of technical debt that makes it way into the code, but I appreciate the point being made here.  This should be a discussion point between the writer and reviewer to agree the best approach.

I think if this scenario happens often it could be a sign that we're not collaborating enough the implementation of a solution before coding so we should see that as a trigger for reviewing our ways of working, rather than blocking PRs.

Matthew-Collins-Defra replied:
Yes, a lot of that depends on the context of the change being performed. If it's the job of the PR to address something, add a new piece of code or extend existing functionality, then requesting a change to include more abstraction or a different interface would be appropriate.
If the change is a small modification, or "temporary" work on the path to something else, then sometimes abstraction can be unwanted.
As we're trying to keep the scope of PRs small, I think the suggestion to add something to the product backlog is a good one. Either it is a priority, it's debt and we need to keep it under control, so the issue will be dealt with as such. If it's not a priority then addressing it in the current PR is inappropriate anyway.
You're right that if this does happen often, then we need to collaborate more to consider the implementation of something before starting. Unless you're doing something new, you frequently know in advance what sort of approach you can take to fix a problem. Stories are small enough that unless the feature that you're working on is very small, the issue that you're fixing will just be a stepping stone on the way to a larger implementation. That will either have further items in the backlog already, or will need them adding as part of this PR.
So trying to solve the entire problem space in one go means that the story is too large and needs to be broken up.

djbingham replied:
Really good point. How about this instead?
> Is there duplication within new code or between new and existing code? Could an existing abstraction be reused or should a new abstraction be created? When proposing an abstraction, consider using a non-blocking comment to open a dialogue about what the abstraction should be and whether it needs addressing now or as a separate piece of work.
>
> Identifying useful abstractions at the review stage may indicate that there wasn't enough collaboration before coding began. See this as a trigger to review the team's ways of working, rather than blocking a pull request unnecessarily.

johnwatson484 replied:
Looks good

Matthew-Collins-Defra replied:
I think that's a good addition.

djbingham replied:
Thanks, I've pushed that.

markysoft replied:
That was the only point that stood out as being contentious, the new wording is better. Loks good to me.

johnwatson484 commented on docs/code-review.md:
```diff
@@ -0,0 +1,97 @@
+3. Block the PR by requesting changes
+
+### When to comment
+You should conclude with a comment when your review asks questions which need answering before you can determine whether the pull request is acceptable. If you haven't proposed a solution to a specific problem in the pull request, it's generally better to leave a neutral review than to block the PR with a request for change.
```
I like this approach, getting an alert saying something has requested changes is always disheartening no matter how thick skinned you are.  

Must confess I often use the requesting changes when only asking questions, when really it would be more appropriate to have just commented.

johnwatson484 reviewed (commented):
Looks good, I think something like this is really important for a team to agree on as early as possible.

I have no concerns with following the approaches documented here.  I've added a few comments to share my opinions on specific areas.  Only issue I found was a slight typo.

Commit:
Correct typo

"seek addition changes" -> "seek additional changes"

Commit:
Add code review link to Readme

Commit:
Clarify that all PR comments should be resolved

Once a comment has been addressed, it should always be marked as
resolved so that everyone can see this is the case.

Commit:
Refine code review standards on duplication

johnwatson484 commented on docs/code-review.md:
```diff
@@ -73,11 +75,11 @@ You should approve a pull request when you have confirmed that:
 Occasionally, it may be prudent to accept a pull request which does not meet all of the above requirements, such as to resolve an urgent issue with the live product. Such cases must always be agreed between the product owner, author(s) and reviewer(s).
 
-If comments or questions on a pull request have been addressed elsewhere (e.g. face-to-face or on Slack), ensure that the outcome is recorded in replies to each comments so that it is visible to anyone looking back at the pull request in future for information.
+If comments or questions on a pull request have been addressed elsewhere (e.g. face-to-face or on Slack), ensure that the outcome is recorded in comment replies so that it is visible to anyone looking back at the pull request in future for information. Use the Resolve button to make it clear which comments have been addressed and which still need attention. A pull request should only be approved  ...
```
Should we clarify who should resolve the conversation?  The writer or the reviewer.

djbingham replied:
That's a great point.

My feeling is that it should be the reviewer so that they explicitly confirm that their concerns have been addressed. It seems too easy for an author to misunderstand a comment and think they've addressed it when they haven't. Sometimes I have resolved comments against my own PRs - when trying to work out what still needed looking at - but it always makes me uncomfortable because of that risk of misunderstanding.

It gets a bit more tricky to know who should resolve it when several reviewers have replied to the same comment thread. Another problem that could occur is a reviewer going on holiday or off sick after leaving comments. I wouldn't want a PR to be held up because the reviewer isn't there to click "Resolve". Perhaps then it should be open to a follow-up reviewer to decide whether the comments have been addressed and resolve them accordingly?

johnwatson484 replied:
I also feel it should be the reviewer but you make a good point about multiple or absent reviewers.

For multiple reviewers, in most cases the conversation will be about the same topic so any of them could resolve as long as the conversation had met a natural conclusion and agreement of what to do and if that involves a change, has that change been made.  But I think it's a judgement call we'd be capable of making.

We could have a follow up reviewer but then that does risk the reviewer also misunderstands the original reviewer.  Do you think we should close it as unresolved so we're not speaking on behalf of the absent reviewer?

djbingham replied:
Yeah, I'm not sure about my idea of a follow-up reviewer hitting resolve either. Sadly, there isn't an option to close a conversation as unresolved, but we could say that it's OK for a follow-up reviewer to approve the pull request with unresolved comments if they think those comments have been addressed and their author is unavailable to confirm.

johnwatson484 replied:
Yeah that's what I meant, close the pull request leaving the comment unresolved.  Should have said "approve" not close.

Commit:
Clarify comment on reusability in review standards

johnwatson484 reviewed (dismissed):
Everyone appears to have reached agreement that this is how we want to approach reviews.

Commit:
Clarify who should resolve conversations on PRs

johnwatson484 reviewed (approved):
LGTM
